gibo: 1.0.6 -> 3.0.14

aileot opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

aileot committed:
gibo: 1.0.6 -> 3.0.14

aileot committed:
gibo: use finalAttrs.version

aileot committed:
gibo: update license

https://github.com/simonwhitaker/gibo/blob/main/LICENSE
https://github.com/simonwhitaker/gibo/blob/v3.0.14/LICENSE
https://github.com/simonwhitaker/gibo/blob/3c71aa988d9b1734a2cd98fc2d6c2612afb1fe7a/LICENSE

LordGrimmauld commented on pkgs/by-name/gi/gibo/package.nix:
```diff
@@ -1,40 +1,50 @@
-        -i "$out/bin/gibo"
-    sed -e 's|\<find |${findutils}/bin/find |g' \
-        -i "$out/share/bash-completion/completions/gibo-completion.bash"
+  installCheckPhase = ''
+    runHook preInstallCheck
+    $out/bin/gibo version | grep -F "${finalAttrs.version}"
+    runHook postInstallCheck
   '';
```
`versionCheckHook` is preferrable over greps.
```suggestion
  doInstallCheck = true;
  nativeInstallCheckInputs = [ versionCheckHook ];
  versionCheckProgramArg = "version" ;
```

LordGrimmauld commented on pkgs/by-name/gi/gibo/package.nix:
```diff
@@ -1,40 +1,50 @@
   meta = {
     homepage = "https://github.com/simonwhitaker/gibo";
-    license = lib.licenses.publicDomain;
+    license = lib.licenses.unlicense;
```
This seems to have been a mistake in the initial package creation, or the (very early) [change to unlicense](https://github.com/simonwhitaker/gibo/commit/6e69c15c811b0508979582d840aec4ed9abb39f8) was missed. But, please link to the actual license (https://github.com/simonwhitaker/gibo/blob/main/LICENSE) in your commit message changing it, makes researching/blaming this change later simpler.

aileot replied:
The commit got  with `git log --grep license` doesn't seem to have any URLs, so I ask, but is the URL to be saved the one to `blob/main` in the commit for the license change?

aileot replied:
I've added the permalink, too.

LordGrimmauld commented on pkgs/by-name/gi/gibo/package.nix:
```diff
@@ -1,40 +1,50 @@
+    license = lib.licenses.unlicense;
     description = "Shell script for easily accessing gitignore boilerplates";
     platforms = lib.platforms.unix;
     mainProgram = "gibo";
```
The 1.0.6 version is *ancient* (Oct 2017). You are likely the first person to care about this package again. You could consider adding yourself to maintainers, as well as adding an update script (`passthru.updateScript = nix-update-script { };`) to make sure this package won't be forgotten about again. Of course, there is no obligation to commit to being a maintainer, just something to consider.

LordGrimmauld replied:
formatter is complaining (needs to be `{ }`, not `{}`), but that does help already.

LordGrimmauld reviewed (changes requested):
Some minor changes that'd be nice, but overall the diff doesn't look bad. I have not yet built the package though.

aileot committed:
gibo: add updateScript

aileot committed:
gibo: use versionCheckHook

Note that `writableTmpDirAsHomeHook` with `$HOME` is necessary to pass
the version check; `gibo` requires either `$XDG_CACHE_HOME` or `$HOME`,
none of which are otherwise defined in running `versionCheckHook`, to
store `gibo`'s boilerplate caches.

https://github.com/NixOS/nixpkgs/pull/422670#pullrequestreview-3021069843

Co-authored-by: Grimmauld <soeren@benjos.de>

LordGrimmauld commented:
(doesn't help the error is just discarded instead of outputting a proper reason: https://github.com/simonwhitaker/gibo/blob/7f070e68b3bdad989f75f9d819cf7af2c1eca5d4/utils/utils.go#L20)

bengsparks reviewed (changes requested):
TL;DR:

```diff
diff --git a/pkgs/by-name/gi/gibo/package.nix b/pkgs/by-name/gi/gibo/package.nix
index 726cc105a434..5c650fa181d2 100644
--- a/pkgs/by-name/gi/gibo/package.nix
+++ b/pkgs/by-name/gi/gibo/package.nix
@@ -2,6 +2,8 @@
   lib,
   buildGoModule,
   fetchFromGitHub,
+  versionCheckHook,
+  writableTmpDirAsHomeHook,
   nix-update-script,
   installShellFiles,
 }:
@@ -36,12 +38,13 @@ buildGoModule (finalAttrs: {
   '';
 
   doInstallCheck = true;
+  nativeInstallCheckInputs = [
+    versionCheckHook
+    writableTmpDirAsHomeHook
+  ];
 
-  installCheckPhase = ''
-    runHook preInstallCheck
-    $out/bin/gibo version | grep -F "${finalAttrs.version}"
-    runHook postInstallCheck
-  '';
+  versionCheckProgramArg = "version";
+  versionCheckKeepEnvironment = [ "HOME" ];
 
   passthru.updateScript = nix-update-script { };
 ```

### Motivation:

The reason for `gibo verison` failing is that Go's `init` routines run at program startup time, which are executed every time, which includes simply trying to look for the version.

One of the routines invoked by a `init` function determines [where gibo stores its boilerplates](https://github.com/simonwhitaker/gibo/blob/7f070e68b3bdad989f75f9d819cf7af2c1eca5d4/cmd/root.go#L18).
In turn, this calls [os.UserCacheDir](https://github.com/simonwhitaker/gibo/blob/7f070e68b3bdad989f75f9d819cf7af2c1eca5d4/utils/utils.go#L18), which looks for `$XDG_CACHE_HOME` and `$HOME` [on Unix and Darwin systems](https://pkg.go.dev/os#UserCacheDir), which do not exist when `versionCheckHook` is running its checks, as it clears its environment by default.

### Solution

Until recently, it was impossible to pass environment variables until https://github.com/NixOS/nixpkgs/commit/23bd27a21c22a84f026cea314d3581344d2f17eb, which implements this via `versionCheckKeepEnvironment`.
To achieve a clean `$HOME` value, we can use `writableTmpDirAsHomeHook`, and simply pass `$HOME` to `versionCheckHook` with `versionCheckKeepEnvironment`.
And voilà, we have our solution, which, I can confirm, builds for both `aarch64-linux` and `aarch64-darwin`.

aileot commented:
@bengsparks Thank you for identifying the cause along with the solution!

bengsparks commented on pkgs/by-name/gi/gibo/package.nix:
```diff
@@ -1,40 +1,57 @@
-        -i "$out/bin/gibo"
-    sed -e 's|\<find |${findutils}/bin/find |g' \
-        -i "$out/share/bash-completion/completions/gibo-completion.bash"
+  postInstall = ''
```
I think this should be, emphasis on the `optionalString` part.
```nix
postInstall = lib.optionalString (stdenv.hostPlatform.emulatorAvailable buildPackages) (
    let
      emulator = stdenv.hostPlatform.emulator buildPackages;
    in
    ''
      installShellCompletion --cmd gibo \
        --bash <(${emulator} $out/bin/gibo completion bash) \
        --fish <(${emulator} $out/bin/gibo completion fish) \
        --zsh <(${emulator} $out/bin/gibo completion zsh)
    ''
  );
  ```

bengsparks reviewed (changes requested):
You'll need to squash the commits into one commit `gibo: 1.0.6 -> 3.0.14`.
You can keep any comments you have in other commits in the latter portion of this final singular commit message, if you so wish.
Otherwise LGTM.

LordGrimmauld commented:
Eh, seperate commits are good, but `gibo: remove redundant version prefix "v"` and `gibo: apply writableTmpDirAsHomeHook with $HOME` should be squashed into update/versionCheckHook respectively

aileot committed:
gibo: enable shell completions only if emulators are available

LordGrimmauld commented on pkgs/by-name/gi/gibo/package.nix:
```diff
@@ -1,40 +1,64 @@
+    installShellFiles
+  ];
+
+  postInstall = lib.optionalString (stdenv.hostPlatform.emulatorAvailable buildPackages) (
+    let
+      emulator = stdenv.hostPlatform.emulator buildPackages;
+    in
+    ''
+      installShellCompletion --cmd gibo \
+        --bash <(${emulator} $out/bin/gibo completion bash) \
+        --fish <(${emulator} $out/bin/gibo completion fish) \
+        --zsh <(${emulator} $out/bin/gibo completion zsh)
+    ''
+  );
```
Emulator is good though, keep that!

LordGrimmauld reviewed (approved):
Alright. Built on `x86_64-linux` and `pkgsCross.aarch64-multiplatform`. Build works, shell completion works, and gibo can successfully update, list and spit out gitignore templates. This seems about good too me.
